README.md
Overview
--------------------
* Name : NOKIA5110_TEXT
* Title : Library for Nokia 5110 LCD (PCD8544 controller) for the Arduino eco-system
* Description : Arduino library,  ASCII character text only (12 * 6 ) 72 characters in total, Sleep mode, minimalist. Single font. Example program(tested on UNO) uses 1874 bytes (5%) of program storage space. Maximum is 32256 bytes.
Global variables use 22 bytes (1%) of dynamic memory, leaving 2026 bytes for local variables. Maximum is 2048 bytes.

* Author: Gavin Lyons
* Arduino 1.8.5

Features
-------------------------
The Nokia 5110 is a basic graphic LCD screen for lots of applications. It was originally intended to be used as a cell phone screen. It uses the PCD8544 controller, which is the same used in the Nokia 3310 LCD. The PCD8544 is a low power CMOS LCD controller/driver, designed to drive a graphic display of 48 rows and 84 columns. All necessary functions for the display are provided in a single chip, including on-chip generation of LCD supply and bias voltages, resulting in a minimum of external components and low power consumption. The PCD8544 interfaces to microcontrollers through a serial bus interface(SPI).

GPIO function on PIC, 5 Nokia 5110 LCD lines SPI bus.

| Arduino    | Nokia 5110 LCD |
| ------ | ------ |
| Digital GPIO | LCD_CLK Pin 5 clock in |
| Digital GPIO | LCD_DIN Pin 4 data in |
| Digital GPIO | LCD_DC Pin 3 data/command|
| Digital GPIO | LCD_CE Pin 2 chip enable |
| Digital GPIO | LCD_RST Pin 1 reset|

Arduino library,  ASCII character text only(12 * 6) 72 characters in total, Sleep mode, minimalist. Single font.

The library uses bit-banging rather than importing SPI libraries.

Example program(tested on UNO) uses 1874 bytes (5%) of program storage space. Maximum is 32256 bytes.
Global variables use 22 bytes (1%) of dynamic memory, leaving 2026 bytes for local variables. Maximum is 2048 bytes.

This compares to Adafruit text based library example of 20% and 38% for program storage and global variables respectively. 

The library has main 3 files (NOKIA5110_TEXT.cpp  NOKIA5110_TEXT.h and NOKIA5110_TEXT_FONT.h).

The screen is 84  X 48 pixels. The X-Axis has 84 columns.
The Y_Axis(rows) the 48 pixels are divided into 6 row blocks. 
Each block containing one byte of pixels. 6 * 8 = 48.

| Block number   | Pixel number Y axis|
| ------ | ------ |
| Block 0 | 0-7 |
| Block 1 | 8-15 |
| Block 2 | 16-23|
| Block 3 | 24-31 |
| Block 4 | 32-39 |
| Block 5 | 40-47 |

The ASCII library in font file: Each character is 5 pixel wide. Plus one pixel of empty space on each side.
Also each character is a byte of pixels in height.
So each character is in a 7 by 8 pixel block. So you can fit (84/7) 12 characters across columns and with (48/8) 6 row blocks that gives 72 characters in total (12X06).

Functions
----------------------------------------

Construct of the class object from left to right pin 1-5(LCD)
RST pin 1, CD pin 2, DC pin 3, DIN pin 4, CLK pin 5  

1. NOKIA5110_TEXT(uint8_t LCD_RST, uint8_t LCD_CE, uint8_t LCD_DC, uint8_t LCD_DIN, uint8_t LCD_CLK)


This sends the  commands to the PCD8544 to init LCD

2. LCDInit()

gotoXY routine to position cursor,  x - range: 0 to 84 (0 to 0x53), 
y - range: 0 to 5 ( 6 blocks one byte each 6 * 8 = 48

3. LCDgotoXY(uint8_t , uint8_t);

Clears the LCD by writing zeros to the entire screen

4. LCDClear(void)

There are two  banks in the LCD, data and commands. 
This function sets the DC pin high or low depending, and then sends
the data byte The first byte is one(data) or zero(cmd) , second byte data

5. LCDWrite(unsigned char , unsigned char);

Given a string of characters, one by one is passed to the LCD

6. LCDString(const char * characters);

Function to set contrast passed a byte , default BF.
 Set LCD VOP Contrast, range = ((0x00-0x7F) |0x80) , 0xB5 = (0x35|0x80) try B1 - BF normally. 

7. LCDsetContrast(uint8_t )

8. LCDenableSleep()
9. LCDdisableSleep()


Function Clears a block(block of rows, one byte) passed block number 0-5 
More efficient than clearing whole screen.

10. LCDClearBlock(uint8_t)


Installation
------------------------------

[Installing Additional Arduino Libraries guide](https://www.arduino.cc/en/Guide/Libraries)

Pictures
---------------------------

![ScreenShot pic](https://github.com/gavinlyonsrepo/NOKIA5110_TEXT/blob/master/extras/image/NOKIA.jpg)

![ScreenShot pic2](https://github.com/gavinlyonsrepo/NOKIA5110_TEXT/blob/master/extras/image/NOKIA2.jpg)




docs/CHANGELOG.md
# Changelog

* version 1.0.0 March 2019
	first release


	

NOKIA5110_TEXT.cpp
/*
* Project Name: Nokia 5110
* File: NOKIA5110_TEXT.cpp
* Description: Nokia library cpp file
* Author: Gavin Lyons.
* IC: ATmega328P 
* IDE:  Arduino 1.8.5
* Created March 2019
* Description: See URL for full details.
* URL: https://github.com/gavinlyonsrepo/NOKIA5110_TEXT
*/

#include "NOKIA5110_TEXT.h"         
#include "NOKIA5110_TEXT_FONT.h"


NOKIA5110_TEXT::NOKIA5110_TEXT(uint8_t LCD_RST, uint8_t LCD_CE, uint8_t LCD_DC, uint8_t LCD_DIN, uint8_t LCD_CLK) {
    
    _LCD_RST = LCD_RST;
    _LCD_CE = LCD_CE;
    _LCD_DC  = LCD_DC;
    _LCD_DIN  = LCD_DIN;
    _LCD_CLK  = LCD_CLK;
  
    pinMode(_LCD_RST,OUTPUT);
    pinMode(_LCD_CE,OUTPUT);
    pinMode(_LCD_DC,OUTPUT);
    pinMode(_LCD_DIN,OUTPUT);
    pinMode(_LCD_CLK,OUTPUT);
}



/*Function : LCDinit
This sends the  commands to the PCD8544 to  init LCD
*/
void NOKIA5110_TEXT::LCDInit(void) {
    //Configure control pins
    _LCD_DIN_SetLow;
    _LCD_CLK_SetLow;
    _LCD_DC_SetLow;
    //Reset the LCD to a known state
    _LCD_RST_SetLow;
    _LCD_RST_SetHigh;
    LCDWrite(LCD_COMMAND, LCD_COMMAND_MODE); //Tell LCD that extended commands follow
    LCDWrite(LCD_COMMAND, LCD_CONTRAST); //Set LCD Vop (Contrast): Try 0xB1(good @ 3.3V) or 0xBF if your display is too dark
    LCDWrite(LCD_COMMAND, LCD_TEMP_COEF); //Set Temp coefficent
    LCDWrite(LCD_COMMAND, LCD_BIAS); //LCD bias mode 1:48: Try 0x13 or 0x14
    LCDWrite(LCD_COMMAND, LCD_FUNCTIONSET ); //We must send 0x20 before modifying the display control mode
    LCDWrite(LCD_COMMAND, LCD_DISPLAYCONTROL); //Set display control, normal mode. 0x0D for inverse
}

/* Function: LCDClear 
Clears the LCD by writing zeros to the entire screen
 */
void NOKIA5110_TEXT::LCDClear(void) {
    for (int index = 0 ; index < (LCD_X * LCD_Y / 8) ; index++)
    LCDWrite(LCD_DATA, 0x00);
    LCDgotoXY(0, 0); //After we clear the display, return to the home position
    }

/* Function: LCDClear 
Clears one of the 6 row blocks(one byte height) on LCD 
by writing zeros to the line. Send the row block number 0-5
 */
void NOKIA5110_TEXT::LCDClearBlock(uint8_t RowBlockNum) {
    LCDgotoXY(0, RowBlockNum);
    for (int index = 0 ; index < (LCD_X) ; index++)
    LCDWrite(LCD_DATA, 0x00);
    }

/* Function: gotoXY gotoXY routine to position cursor 
 x - range: 0 to 84 (0 to 0x53)
 y - range: 0 to 5 ( 6 blocks one byte each 6*8 = 48*/
void NOKIA5110_TEXT::LCDgotoXY(uint8_t  x, uint8_t  y) {
    LCDWrite(LCD_COMMAND, 0x80 | x); // Column. (result 0x80 to 0xD3)
    LCDWrite(LCD_COMMAND, 0x40 | y); // Row.
}

/* Function: LCDWrite 
There are two  banks in the LCD, data and commands. This
function sets the DC pin high or low depending, and then sends
the data byte
 */
void NOKIA5110_TEXT::LCDWrite(unsigned char data_or_command, unsigned char data) {
    unsigned char i,d;
    d=data;
    if(data_or_command==0)_LCD_DC_SetLow;
    else _LCD_DC_SetHigh;
    //data_or_command; 
    //Tell the LCD that we are writing either to data or a command
    //Send the data
    _LCD_CE_SetLow;
    for(i=0;i<8;i++)
    {
        _LCD_DIN_SetLow;
        if(d&0x80)_LCD_DIN_SetHigh; // b1000000 Mask with 0 & all zeros out.
        _LCD_CLK_SetHigh;
        d<<=1;
        _LCD_CLK_SetLow;
    }
   _LCD_CE_SetHigh;
}

/* Function: LCDCharacter.
 This function takes in a character, looks it up in the font table/array
And writes it to the screen
Each character is 8 bits tall and 5 bits wide. We pad one blank column of
pixels on each side of the character for readability.
 */
void NOKIA5110_TEXT::LCDCharacter(char character) {
    LCDWrite(LCD_DATA, 0x00); //Blank vertical line padding
    for (uint8_t index = 0 ; index < 5 ; index++)
    {
    LCDWrite(LCD_DATA, (pgm_read_byte(&ASCII[character - 0x20][index])));
    }
    // LCDWrite(LCD_DATA, ASCII[character - 0x20][index]);
    //0x20 is the ASCII character for Space The font table starts with this character
    LCDWrite(LCD_DATA, 0x00); //Blank vertical line padding
}

/* Function: LCDString.
 Given a string of characters, one by one is passed to the LCD
 */
void NOKIA5110_TEXT::LCDString(const char  *characters) {
    while (*characters)
    LCDCharacter(*characters++);
}


/* Function: LCDSetContrast
 Function to set contrast passed a byte 
 Set LCD VOP Contrast, range = ((0x00-0x7F) |0x80) 0xB5 = (0x35|0x80) try B1 - BF normally. 
 */
void NOKIA5110_TEXT::LCDsetContrast(uint8_t contrast)
{
    _contrast = contrast;
    LCDWrite(LCD_COMMAND, LCD_COMMAND_MODE); //Tell LCD that extended commands follow
    LCDWrite(LCD_COMMAND, contrast); //Set LCD Vop (Contrast): 
    LCDWrite(LCD_COMMAND, LCD_FUNCTIONSET ); //We must send 0x20 before modifying the display control mode
}

// Put the LCD to Sleep function
void NOKIA5110_TEXT::LCDenableSleep()
{
    _sleep = true;
     for (int index = 0 ; index < (LCD_X * LCD_Y / 8) ; index++)
    LCDWrite(LCD_DATA, 0x00);
    LCDgotoXY(0, 0);
    LCDWrite(LCD_COMMAND, LCD_POWERDOWN);
}

// Wake the LCD up from Sleep function
void NOKIA5110_TEXT::LCDdisableSleep()
{
   
    _sleep = false;
    LCDWrite(LCD_COMMAND , LCD_COMMAND_MODE);
    LCDWrite(LCD_COMMAND , _contrast);
    LCDWrite(LCD_COMMAND,  LCD_TEMP_COEF );
    LCDWrite(LCD_COMMAND,  LCD_BIAS );
    LCDWrite(LCD_COMMAND, LCD_FUNCTIONSET );
    LCDWrite(LCD_COMMAND, LCD_DISPLAYCONTROL); 
}

/* =========== EOF ===========*/